refactor(sdk): split Rust SDK into smaller packages

marvin-hansen opened the pull request:
As discussed in issue  https://github.com/apache/iggy/issues/1713

This PR slims down the SDK crate by a good chunk, creates a new binary protocol crate,
and tidies up a few things.

Achieved:
* Smaller SDK crate that is simpler to reason about
* Consolidate and clean up shared code 
* Isolate and version the binary protocol in a dedicated crate to structure compatibility

New crates:
* common (iggy_common)
* binary_protocol (iggy_binary_protocol)

Not achieved:
- Types and errors could not be disentangled thus all of them moved in the  iggy_common crate
- network transport has proven to be equally entangled so its still in the SDK crate. 

Net change:
Before: SDK was about ~33K loc
After: SDK is about ~15k Loc

Solid 50% reduction. Not as much as initially hoped, but way better than before.

spetz commented:
Thank you for such a great contribution! It looks like it doesn't build locally due to missing `s3` dependency used e.g. by `archiver` (could be removed by `cargo machete` or so).

I was thinking if the `messaging` module from `common/messages` could be moved one level above and maybe merged with the existing `messages_shared` (maybe even renamed afterwards to `messages`)?

Most of the modules under `messages` seem to be grouping the requests/commands, thus I was also wondering if it would make sense to call these `commands` instead, or maybe even move all these directories out of the `messages` parent directory?

Naming and grouping all these types is challenging, but looks like, all the hard work is already done :)

spetz commented on core/binary_protocol/src/binary_clients/binary_client.rs:
```diff
@@ -0,0 +1,7 @@
+use crate::binary_clients::client::Client;
```
Looks like this one is missing ASF header.

numinnex commented:
I like the general idea, a few things that I would do differently is moving the `messaging` module from `messages` one level above into `types`.  Other than that I would move `flush_unsaved_buffer` `poll_messages` and `send_messages` into a dedicated module (maybe even call it `messaging`) inside of `messages`. Those are the two things that I feel like are inconsistent with the rest of the arrangement.

marvin-hansen commented:
Thank you @spetz @numinnex 

I've updated the PR:
* added s3 back (was removed by machete)
* added the missing ASF header
* Merged messaging with messages_shared and moved it into types.
* Moved the CLI protocol from SDK into the binary_protocol crate.

The last one also makes the CLI feature flag in the SDK crate obsolete, so I removed it for now. 

As of now, the SDK crate is down to a bit over 8k LoC.

Please take another look.

Thank you

spetz commented:
Thanks for updates! Everything now builds correctly, and all the tests pass too.

Speaking of the structure, some of the commands such as `PollMessages`, `SendMessages` etc. were moved to to `types/message` while all the remaining commands are under `commands` directory. I was thinking, if we should either move these structs to e.g. `commands/messaging`, or, as suggested previously, we could remove the `commands` at all, and move all its submodules into the main level, next to the existing modules such as `traits`, `locking` and so on?

TBH not sure, what approach would make more sense.

marvin-hansen commented:
As discussed on Discord, moving the lost commands from types back into commands/messaging

spetz commented:
Thank you for all the improvements, great work, LGTM!

numinnex commented:
From the structural side LGTM. We can merge this PR, there are still some things that I would like to tackle in another occasion, for example we re-export as part of our prelude certain types such as `IggyIndexView`, which means that user of our sdk can do something like this - `let idx_view = IggyIndexView::new(&[0;16]);`. It's not something terribly bad, but unnecessary, I think those types that are for our own internal usage, shouldn't be publicly available, but as I've said that's something we can take care of in other PR.

marvin-hansen commented:
Okay great. Thanks everyone. 
Someone else has to merge as I don't have permission to do so.
Have a fun weekend!
